### ISSUE TITLE
Travis: check indentation tabs in .sc/.scd files

### ISSUE THREAD

**Issue by crucialfelix** [supercollider/supercollider#2094]
This would catch indentation errors in pull requests (and fail fast).

The same PR that solves this would have to also fix any current problem files.

**Comment by crucialfelix** [comment]
https://github.com/evanshortiss/lintspaces-cli

**Comment by timblechmann** [comment]
tabs in sclang/scsynth/plugins, spaces in supernova/sc-ide (spaces are used in almost any modern coding style like google, llvm, mozilla, webkit, qt).

maybe sclang/scsynth/plugins can/should be converted to spaces as well. less headache in the long run and tabs only make sense when defining the number of spaces per tabs (compare linux kernel guidelines)

**Comment by crucialfelix** [comment]
oh right it is actually tabs in sclang/scsynth/plugins C code. I forgot.  I would support going to spaces for that.

.sc / .scd code should stay tabs else it would traumatize everybody.

**Comment by muellmusik** [comment]
> On 17 May 2016, at 08:23, Chris Sattinger <notifications@github.com> wrote:
> 
> oh right it is actually tabs in sclang/scsynth/plugins C code. I forgot. I would support going to spaces for that.
> 
Is there a way to do that without hosing git blame for every indented line? That will be a major pain.

S.

**Comment by crucialfelix** [comment]
true, that would be a bit much.

**Comment by timblechmann** [comment]
>> oh right it is actually tabs in sclang/scsynth/plugins C code. I
> forgot. I would support going to spaces for that.
>>
> Is there a way to do that without hosing git blame for every indented
> line? That will be a major pain.

git blame -w

> Ignore whitespace when comparing the parent's version and the child's to find where the lines came from.

one could also rewrite the complete history via git-filter-branch. i
personally would not recommend going this direction, as it will change
all git hashes, invalidating all checkouts/clones/forks.

might scare the shit out of users working on a local checkout. i'd
recommend to recommend 'git blame -w'.

**Comment by gusano** [comment]
> maybe sclang/scsynth/plugins can/should be converted to spaces as well

oh yes :+1:

**Comment by miguel-negrao** [comment]
> maybe sclang/scsynth/plugins can/should be converted to spaces as well
> 
> oh yes :+1:

:+1:  !!!!

When working in the projects feature it was incredibly tedious to change the tab/spaces settings every time I changed from sclang code to ide code and back... the less tedious it is to work on SC, the more people will contribute...

**Comment by gusano** [comment]
> .sc / .scd code should stay tabs else it would traumatize everybody

@crucialfelix Going for spaces everywhere *but* the library would be bad IMHO. If we are to change it and go the-modern-way why not doing it for everything at once?

**Comment by telephon** [comment]
? so far, everything is tabs, not spaces. Why not convert everything to tabs?

**Comment by crucialfelix** [comment]
@gusano you are a professional developer. you have to go ask some normal people what they think.

besides, then we would have to have the 4 versus 2 argument. and whoever argues the longest wins.

**Comment by miguel-negrao** [comment]
> ? so far, everything is tabs, not spaces. Why not convert everything to tabs?

Here we go, [the tab vs spaces debate](https://www.google.com/search?q=the+tabs+vs+spaces+debate), it's endless fun. :-)

**Comment by gusano** [comment]
Guys, I didn't mean to start another tabs/spaces war, I thought there was a sudden consensus about moving to spaces.. Sorry for the noise! :japanese_ogre:

**Comment by muellmusik** [comment]
Actually @telephon has a point. The original code was tabs, and sclang/scsynth/plugins/classlib are still for the most part tabs. The problem started when people started adding code that did not respect the conventions of the project.

If there's going to be a conversion, it would be less invasive to convert the space files to tabs.

If (as I suspect) this will end in a lack of consensus, then perhaps the middle ground would at least be to formalise spaces in supernova/ide, tabs everywhere else. That's perhaps not ideal, but at least it's fairly easy to understand.

**Comment by telephon** [comment]
I somehow enjoyed my first attempt at trolling!

**Comment by muellmusik** [comment]
> I somehow enjoyed my first attempt at trolling!

LMAO!

**Comment by timblechmann** [comment]
yes, usually the people shouting loudest will win. sc's codebase is aged, attempts to modernize it are usually facing resistance. old codebase also isn't exactly attractive for new developers to get into, either.

happy flaming.

**Comment by crucialfelix** [comment]
Of course this ticket is about setting up travis to check PRs against the current standard. If somebody wants to propose switching proposals then make your own issue ;)

**Comment by crucialfelix** [comment]
So https://github.com/evanshortiss/lintspaces-cli would install and run quickly on travis, AND it can use an [.editorconfig](http://editorconfig.org/) file !

Which means you can [install a plugin](http://editorconfig.org/#download) and your IDE automatically switches to tabs/spaces based on your spec. Many of them fix on save. Or code beautifiers can use it.

This is my .editorconfig for sc. It needs a slight fix for the cpp but otherwise I can make a PR for it.

```
# EditorConfig is awesome: http://EditorConfig.org

# top-most EditorConfig file
root = true

# Unix-style newlines with a newline ending every file
[*]
end_of_line = lf
insert_final_newline = true

[*.{h,hpp,cpp}]
indent_style = space
indent_size = 4

# Tab indentation
[*.{sc,schelp}]
indent_style = tab
indent_size = 4

[CMakeLists.txt]
indent_style = tab
indent_size = 4

# Matches the exact files either .travis.yml
[*.{yml,yaml,json,js}}]
indent_style = space
indent_size = 2
```

**Comment by crucialfelix** [comment]
billions of errors

```
lintspaces -e .editorconfig ./lang/**/*.cpp
lintspaces -e .editorconfig ./include/**/*.h
```

**Comment by muellmusik** [comment]
> happy flaming.

Flaming? This has got to be the most good-humoured conversation about tabs vs spaces ever! ;-)

**Comment by miguel-negrao** [comment]
> Which means you can install a plugin and your IDE automatically switches to tabs/spaces based on your spec. Many of them fix on save. Or code beautifiers can use it.
> 
> This is my .editorconfig for sc. It needs a slight fix for the cpp but otherwise I can make a PR for it.

But can editorconfig have different settings for different folders ? The settings you posted above have the same settings for all .cpp files, that is not going to work since ide is spaces and sclang, etc is tabs, right ? Also, a pity that QtCreator doesn't support editorconfig apparently, although it has [its own](http://doc.qt.io/qtcreator/creator-sharing-project-settings.html) shareable settings file type. I don't think it is possible to have QtCreator use different settings for different folders through...

**Comment by gusano** [comment]
> But can editorconfig have different settings for different folders ?

@miguel-negrao Yes:

```
[editors/sc-ide/*/**.cpp]
indent_style = space
indent_size = 4

[lang/*/**.cpp]
indent_style = tab
indent_size = 4
```

**Comment by miguel-negrao** [comment]
> But can editorconfig have different settings for different folders ?
> 
> @miguel-negrao Yes:
> 
> [editors/sc-ide/*/**.cpp]
> indent_style = space
> indent_size = 4
> 
> [lang/*/**.cpp]
> indent_style = tab
> indent_size = 4

Great ! Also, it's not on the website, but on [github](https://github.com/editorconfig/editorconfig-qtcreator) they have a plugin for QtCreator, although I'm not sure if it will respect the settings for different folders...

**Comment by crucialfelix** [comment]
CMakeLists.txt should be all spaces.  Only a few are tabs, some are mixed. 

I will post a PR with the .editorconfig

**Comment by crucialfelix** [comment]
This is better: https://github.com/jedmao/eclint

It can also fix files automatically.

**Comment by miczac** [comment]
On 17.May 2016, at 22:50 , muellmusik <notifications@github.com> wrote:

> perhaps the middle ground would at least be to formalise spaces in supernova/ide, tabs everywhere else. That's perhaps not ideal, but at least it's fairly easy to understand.

maybe formalise/convert to spaces in supernova/scsynth/ide as in for �in all the binaries� ?

**Comment by brianlheim** [comment]
Updating this issue now that #2819 is closed; we now lint for spaces in all C++ files; only SC code is unlinted.

**Comment by dyfer** [comment]
We are linting SC code now in GHA, but it's not yet enforce (i.e. the run does not fail).
